scripts: gen_defines.py: Add out_dev() flag for deprecated macros

ulfalizer opened the pull request:
Add a 'deprecation_msg' string/flag to out_dev(). When 'deprecation_msg'
is passed, all generated macros include

    __WARN(<deprecation_msg>)

which prints a custom warning if the macro is used.

Meant to be used when improving the output format.

ulfalizer commented:
This commit also appears in https://github.com/zephyrproject-rtos/zephyr/pull/21506. Adding it separately in case that PR gets stalled.

galak commented on scripts/dts/gen_defines.py:
```diff
@@ -771,6 +776,17 @@ def out(ident, val, aliases=()):
+
+    s = "#define DT_{:40}".format(ident)
+    if deprecation_msg:
+        s += r' _Pragma("GCC warning \"{}\"")'.format(deprecation_msg)
```
Can we have this be __DEPRECATED_MACRO instead.  So that way this isn't GCC specific.

galak reviewed (changes requested):
Make what we produce in the define use `__DEPRECATED_MACRO` so its not GCC specific.

galak commented:
looks like we need to introduce a `__DEPRECATED_MACRO_MSG` that allows passing a message.

ulfalizer committed:
scripts: gen_defines.py: Add out_dev() flag for deprecated macros

Add a 'deprecation_msg' string/flag to out_dev(). When 'deprecation_msg'
is passed, all generated macros include

    __WARN(<deprecation_msg>)

which prints a custom warning if the macro is used.

Meant to be used when improving the output format.

Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

ulfalizer commented:
Rebased now. Depends on https://github.com/zephyrproject-rtos/zephyr/pull/22336, so marking DNM until that one is in.

galak commented:
Removing DNM now that #22336 is merged.
